internal/ratelimit/monitor.go

```diff
@@ -167,9 +167,13 @@ func (c *Monitor) RecommendedWaitForBackgroundOp(cost int) (timeRemaining time.D
 // It returns true if rate limiting was applying, and false if not.
 // This can be used to determine whether or not a request should be retried.
 func (c *Monitor) WaitForRateLimit(ctx context.Context) bool {
+	// NOTE: We manually unlock the mutex in this function so that it does not remain locked while sleeping
+	c.mu.Lock()
```

Comment:
I think it would be safer to unlock in a defer statement, that way we should capture all escape paths from this function with an unlocked mutex:
```suggestion
	c.mu.Lock()
	defer c.mu.Unlock()
```

Reply:
We cannot defer an unlock because then it will remain locked while sleeping

Reply:
So, do we do that or not? 
<img width="432" alt="screenshot_2023-03-07_11 17 38@2x" src="https://user-images.githubusercontent.com/1185253/223393159-0e103135-790d-4089-9618-23cead9d1cda.png">

😄

Reply:
So I ended up working around this by separating the duration calculation logic to a new function. Then we can defer inside that function, but the actual sleep happens in the calling function.